### PR TITLE
Add domain decomp

### PR DESCRIPTION
**Pull request by arporter**
This is for work done in #5. It adds an option to build the library with MPI support and generates a domain decomposition for the grid.

Rather than make MPI a mandatory dependency, I've implemented two modules, `parallel_mod.f90` and `parallel_stub_mod.f90` that provide the same interfaces. The build system creates a link to one or other of these files, depending on whether DM support is required, prior to the compilation.

### CODE REVIEW

**Comment by arporter** [comment]
@hiker - this is ready for review now if you'd be so kind. I can't explicitly request a review from you yet because you weren't a collaborator on this project - you should have an invitation for that now.

File: finite_difference/src/subdomain_mod.f90
Comment (hiker) [inline finite_difference/src/subdomain_mod.f90]: Wouldn't it be more descriptive to use type(region_type) :: external instead of separate {x,y}{start,stop}/n{x,y}?
This would make it obvious everywhere whether you are using the external or internal domain.

File: finite_difference/src/subdomain_mod.f90
Comment (hiker) [inline finite_difference/src/subdomain_mod.f90]: The comment must go before the function, otherwise doxygen does not pick it up.

File: finite_difference/src/subdomain_mod.f90
Comment (hiker) [inline finite_difference/src/subdomain_mod.f90]: Would it potentially be useful to support different halo widths in east-west and north/south direction at least? Considering that we had a long discussion on how to specify various stencils in PSyclone ;)
Saying that, I am happy to just accept one halo size ... since this will likely cover 95% of all use cases ;)

File: finite_difference/src/subdomain_mod.f90
Comment (hiker) [inline finite_difference/src/subdomain_mod.f90:157]: I would imagine that there are some use cases for distributed fields without halo - e.g. if you just compute an output field from other fields (likely with halo). So it might be too restrictive to enforce those fields to have a halo.

File: finite_difference/src/subdomain_mod.f90
Comment (hiker) [inline finite_difference/src/subdomain_mod.f90:196]: Impressive attention to detail :+1: I would have forced the user to supply the domain decomposition, but this is certainly useful - well done!

File: finite_difference/src/subdomain_mod.f90
Comment (hiker) [inline finite_difference/src/subdomain_mod.f90]: Please remove commented out code.

File: finite_difference/src/subdomain_mod.f90
Comment (hiker) [inline finite_difference/src/subdomain_mod.f90]: This seems to me to be way too complicated - instead of rounding just always round down. This way you have only two cases to consider (junder=0 or junder>0), which should simplify the code further down as well, and I can't see any disadvantage in this approach.

File: finite_difference/src/subdomain_mod.f90
Comment (hiker) [inline finite_difference/src/subdomain_mod.f90]: Remove commented out code

File: finite_difference/src/subdomain_mod.f90
Comment (hiker) [inline finite_difference/src/subdomain_mod.f90]: Ditto for rounding.

File: finite_difference/src/subdomain_mod.f90
Comment (hiker) [inline finite_difference/src/subdomain_mod.f90]: Remove - not used here, and it is being redefined later.

File: finite_difference/src/subdomain_mod.f90
Comment (hiker) [inline finite_difference/src/subdomain_mod.f90:260]: With rounding down in line 147 one else branch would be unnecessary.

File: finite_difference/src/subdomain_mod.f90
Comment (hiker) [inline finite_difference/src/subdomain_mod.f90]: Isn't ...%nx = width? That looks a lot simpler ;)

File: finite_difference/src/subdomain_mod.f90
Comment (hiker) [inline finite_difference/src/subdomain_mod.f90]: ny = height

**Review by hiker — COMMENTED** [review]
Review of the remaining code still to be done by me.

File: finite_difference/src/grid_mod.f90
Comment (hiker) [inline finite_difference/src/grid_mod.f90:4]: Add a doxygen comment "!>" at beginning of file, otherwise it appears no doxygen code is created.

File: finite_difference/src/grid_mod.f90
Comment (hiker) [inline finite_difference/src/grid_mod.f90:130]: While it might be unrelated to this pull request - but why a two step initialisation process? Shouldn't one constructor be enough?

File: finite_difference/src/grid_mod.f90
Comment (hiker) [inline finite_difference/src/grid_mod.f90]: I am not sure I understand the tmask handling here: the fact that a tmask is present is not used in the code at all, the two if and else blocks appear to be identical, except for the abort statement in case of PBC and >1 MPI rank.

I could imagine code that do not use the tmask concept at all, but still want to use an DM parallelisation, which would not be possible this way.

File: finite_difference/src/argument_mod.F90
Comment (hiker) [inline finite_difference/src/argument_mod.F90]: This seems to be unused at this stage? I don't know how this is going to be used, but that's around 5 (to 10 in case of 8 byte default integer) :)

File: finite_difference/src/field_mod.f90
Comment (hiker) [inline finite_difference/src/field_mod.f90:193]: Not sure if I get this: supporting tiling on an OpenMP level seems to be a good idea, but the constructor of the grid allocated tmask and other fields without the tiling. So potentially you would (unnecessarily) access remote memory when mixing tmask (etc), and fields allocated here.
Wouldn't it make more sense to:
1) compute the thread tiling in the grid constructor (or init function)
2) Then use the the field type to create the tmask (etc) variables that are part of the grid object.
This would make this function somewhat cheaper (since all this reading of environment variable etc is done only once).

That might be a big change, and I am happy to accept the code and discuss this in a separate ticket.

File: finite_difference/src/field_mod.f90
Comment (hiker) [inline finite_difference/src/field_mod.f90]: Apparently I can't comment on line 242 - but if we support tiling as indicated above, then the openmp loop in line 242 should use the tiling information. Plus, as far as I can tell: this would also need a change in psyclone ... I am getting the feeling I am missing something :)

File: finite_difference/src/parallel/parallel_stub_mod.f90
Comment (hiker) [inline finite_difference/src/parallel/parallel_stub_mod.f90]: Not sure about coding style: do we allow generic use? I would prefer:
use parallel_common_mod, only : rank, nranks, nprocx, nprocy

File: finite_difference/src/parallel/parallel_mod.f90
Comment (hiker) [inline finite_difference/src/parallel/parallel_mod.f90]: Docstring.

File: finite_difference/src/parallel/parallel_mod.f90
Comment (hiker) [inline finite_difference/src/parallel/parallel_mod.f90]: Docstring.

**Review by hiker — CHANGES_REQUESTED** [review]
Good work. I am not sure about a few of my comments (esp. re tiling). Also I noticed quite a few modules would not create docygen, and while I initially thought it was related to missing !> at the top of a file (there is a comment somewhere), this seems not to be the right fix in other cases.

File: finite_difference/src/subdomain_mod.f90
Comment (arporter) [reply finite_difference/src/subdomain_mod.f90]: That's a good idea. I've used "global" rather than "external" as it holds information on where the subdomain sits within the global one.

File: finite_difference/src/subdomain_mod.f90
Comment (arporter) [reply finite_difference/src/subdomain_mod.f90]: Done.

File: finite_difference/src/subdomain_mod.f90
Comment (arporter) [reply finite_difference/src/subdomain_mod.f90]: I think that could be a future addition - I can imagine adding an optional `halo_widths(4)` argument to this routine. I think the rest of the library has enough flexibility to support this feature in the future.

File: finite_difference/src/subdomain_mod.f90
Comment (arporter) [reply finite_difference/src/subdomain_mod.f90:157]: Oh yes, good point. However, this routine computes a general decomposition, independent of any particular field. This might be influenced by some of your points below though so I'll hold off for now.

File: finite_difference/src/subdomain_mod.f90
Comment (arporter) [reply finite_difference/src/subdomain_mod.f90:196]: Thanks :-) It was some code I had "prepared earlier" (do you have Blue Peter in Oz?)

File: finite_difference/src/subdomain_mod.f90
Comment (arporter) [reply finite_difference/src/subdomain_mod.f90]: Well spotted - done.

File: finite_difference/src/subdomain_mod.f90
Comment (arporter) [reply finite_difference/src/subdomain_mod.f90]: Oh yes. Done.

File: finite_difference/src/subdomain_mod.f90
Comment (arporter) [reply finite_difference/src/subdomain_mod.f90]: Done.

File: finite_difference/src/subdomain_mod.f90
Comment (arporter) [reply finite_difference/src/subdomain_mod.f90]: True but the compiler does not know that `ntilex` is guaranteed to be > 1 (and thus that the `ji` loop will do at least one trip) and therefore warns about the assignment using the `subdomain` pointer after the loop over `ji`. This is a simple way of preventing that warning. I've extended the associated comment to explain this.

File: finite_difference/src/subdomain_mod.f90
Comment (arporter) [reply finite_difference/src/subdomain_mod.f90:260]: Done.

File: finite_difference/src/subdomain_mod.f90
Comment (arporter) [reply finite_difference/src/subdomain_mod.f90]: Oh yes. Ahem. :-)

File: finite_difference/src/subdomain_mod.f90
Comment (arporter) [reply finite_difference/src/subdomain_mod.f90]: Done.

File: finite_difference/src/grid_mod.f90
Comment (arporter) [reply finite_difference/src/grid_mod.f90:130]: I think I did it this way so that you can pass the grid object down to other routines that e.g. read from file in order to set-up the T mask.

File: finite_difference/src/grid_mod.f90
Comment (arporter) [reply finite_difference/src/grid_mod.f90]: Ah! The benefits of code review. When the code was serial I used the T-mask to set-up the size of the grid. Now however I use the domain decomposition so you are right - the if block was completely unecessary.

Comment (arporter) [reply finite_difference/src/grid_mod.f90]: Support for the case where we have no T mask and are running in parallel is something for a future issue...

File: finite_difference/src/argument_mod.F90
Comment (arporter) [reply finite_difference/src/argument_mod.F90]: It's needed in order to make any stencil meta-data in a kernel valid Fortran. I too am concerned about the size but hopefully the compiler will optimise it out as the quantities concerned are never actually used.

File: finite_difference/src/field_mod.f90
Comment (arporter) [reply finite_difference/src/field_mod.f90:193]: You  are right again - I did begin to think that it was odd that I'd (originally) chosen to set-up the thread tiling repeatedly for each new field that was created. I've created #8 for this.

File: finite_difference/src/field_mod.f90
Comment (arporter) [reply finite_difference/src/field_mod.f90]: Oh yes, it should - done. I think PSyclone should be OK for now but might require changing if we do change the way fields have tiles.

File: finite_difference/src/argument_mod.F90
Comment (rupertford) [reply finite_difference/src/argument_mod.F90]: I don't think this is a reasonable way to go, particularly if we are going to support larger stencils (potentially requiring up to 0:999, 0:919, 0:999). I have a solution that uses structure constructors so does not need an array at all. It works with gfortran so I assume it is OK. I am using this in the manual ompss implementation of nemolite2d. Perhaps you could create an issue and I could add this solution in?

File: finite_difference/src/parallel/parallel_stub_mod.f90
Comment (arporter) [reply finite_difference/src/parallel/parallel_stub_mod.f90]: Done.

File: finite_difference/src/parallel/parallel_mod.f90
Comment (arporter) [reply finite_difference/src/parallel/parallel_mod.f90]: Well spotted. Done - and below.

**Comment by arporter** [comment]
I've added a configuration file and build target for doxygen-generated documentation.
I think I've addressed all comments.
Ready for second review...

File: finite_difference/src/argument_mod.F90
Comment (hiker) [reply finite_difference/src/argument_mod.F90]: Great :)

File: finite_difference/src/field_mod.f90
Comment (hiker) [reply finite_difference/src/field_mod.f90:193]: Imho this should be tied in with PSyclone issue 109 (https://github.com/stfc/PSyclone/issues/109) - ideally this runtime library should provide information to PSyclone on what loops to generate. Then installing the runtime library and using it with PSyclone would created either a tiled or nested loops - but that's far far out of scope of this - am happy with the ticket.

File: finite_difference/src/tile_mod.f90
Comment (hiker) [inline finite_difference/src/tile_mod.f90:29]: I'd suggest that all new files at least have the doxygen module line to have the modules listed (e.g. in finite_difference/src/doc/html/annotated.html):
```
!> Module containing the definition of the tile type
```

File: finite_difference/src/grid_mod.f90
Comment (hiker) [reply finite_difference/src/grid_mod.f90:130]: Makes sense, thanks for the explanation.

File: finite_difference/src/subdomain_mod.f90
Comment (hiker) [reply finite_difference/src/subdomain_mod.f90:196]: Nope :)

File: finite_difference/src/dl_esm_inf.doxygen
Comment (hiker) [inline finite_difference/src/dl_esm_inf.doxygen]: Can you rename this config file to Doxyfile or doxyfile? That's the default, allowing us to run doxygen without specifying a filename.

File: finite_difference/src/grid_mod.f90
Comment (hiker) [reply finite_difference/src/grid_mod.f90:4]: Still most modules have the module information missing (e.g. in finite_difference/src/doc/html/annotated.html), but I'll only request new files to add them (and have added appropriate comments elsewhere).

File: finite_difference/src/parallel_common_mod.f90
Comment (hiker) [inline finite_difference/src/parallel_common_mod.f90]: Could you move those two comments lines to before the module line, then the module will be listed in doxygen.

**Review by hiker — CHANGES_REQUESTED** [review]
A few changes only, mostly (exclusively??) related to doxygen.

File: finite_difference/src/subdomain_mod.f90
Comment (arporter) [reply finite_difference/src/subdomain_mod.f90:196]: Oh well. Kid's TV show where they do a craft activity and then skip to the finished article with a "here's one I prepared earlier".

File: finite_difference/src/tile_mod.f90
Comment (arporter) [reply finite_difference/src/tile_mod.f90:29]: Done.

File: finite_difference/src/dl_esm_inf.doxygen
Comment (arporter) [reply finite_difference/src/dl_esm_inf.doxygen]: Done.

File: finite_difference/src/parallel_common_mod.f90
Comment (arporter) [reply finite_difference/src/parallel_common_mod.f90]: Done.

**Comment by arporter** [comment]
Doxygen does now generate the full list of modules which makes me feel good :-)

Back to @hiker for 3rd(?) review...

**Review by hiker — APPROVED** [review]

**Comment by hiker** [comment]
All looking good now. My above comment that some files are missing in ... annotated.html is incorrect (it apparently lists only public types, so files without a public type declaration will not be listed there).

**Comment by hiker** [comment]
Can you just bring it up-to-date with master (I probably merged in the wrong order.? Sorry)

**Pull request merged**
